### ISSUE TITLE
Bug: base class expression infer error

### ISSUE THREAD

**Issue by yanyongyu** [microsoft/pyright#5326]
**Describe the bug**

I want to create some fake class in testing to override the base class. So, ...

1. Create a dynamic class as shown in bellow.

Pylance raises two errors:

- `Expected type expression but received "Type[A] | None"` on var `base`
- `Argument to class must be a base class` on expression `base or A`

Note that if i create a var equals to `base or A` and then inherit it, all is OK.

**To Reproduce**

```python
from typing import Optional, Type


class A:
    ...

def make_fake_class(base: Optional[Type[A]] = None):
    class FakeClass(base or A):  # Errors: Expected type expression but received "Type[A] | None"; Argument to class must be a base class
        ...
    
    Base = base or A
    
    class FakeEvent(Base):
        ...

    return FakeClass
```

**Expected behavior**

No error should be raised

**Code or Screenshots**

![image](https://github.com/microsoft/pyright/assets/42488585/39de735a-9edf-4d4f-862a-52be2b80e6ab)

**VS Code extension or command-line**

-   Language Server version: Pylance language server 2023.6.20 (pyright 73165c50)
-   OS and version: Linux
-   Python version (& distribution if applicable, e.g. Anaconda): codespaces 3.10 venv

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.315, which I just published. It will also be included in this week's insiders release of pylance.
